src/state/plugin.ts

```diff
@@ -0,0 +1,45 @@
+import type MarkdownIt from "markdown-it"
+import { RuleCore } from "markdown-it/lib/parser_core"
+import type StateCore from "markdown-it/lib/rules_core/state_core"
+import { getDocState } from "./utils"
+
+/** Allowed options for state plugin */
+export interface IOptions {
+  // TODO: Figure out options
+  autonumberFigures: boolean
+}
+
+function numberingRule(options: IOptions): RuleCore {
+  return (state: StateCore) => {
+    const env = getDocState(state)
+
+    env.references.forEach(ref => {
+      const { name, tokens, contentFromTarget } = ref
+
+      function setError(details: string) {
+        tokens.open.attrJoin("class", "error")
+        tokens.open.tag = tokens.close.tag = "span"
+        tokens.content.content = details
+        return true
+      }
+
+      const target = env.targets[name]
+      if (!target) {
+        return setError(`Reference "${name}" not found`)
```

Comment:
Yeh I feel in something like jupyterlab-myst, where I believe we are currently only passing single cells at a time, there could be a lot of "un-matched" references (because their target is somewhere else on the page, or a cross-page ref). So then it would probably be nicer to just have the content as only the `${name}`, and let the CSS for the error class "indicate" that the reference was not found